### PR TITLE
St7735

### PR DESCRIPTION
**Pull request by Experiment-626**
Preliminary ST7735 driver.

### CODE REVIEW

**Comment by Experiment-626** [comment]
The low level WriteData function was never called by the higher level WriteData functions (now SendData). This would mean that if the user sent a draw buffer via SendData it would be very slow.

File: Sitronix/GHIElectronics.TinyCLR.Drivers.Sitronix.ST7735/ST7735.cs [hunk 1]
```diff
@@ -1,5 +1,222 @@
-﻿namespace GHIElectronics.TinyCLR.Drivers.Sitronix.ST7735 {
+﻿using System.Threading;
+
+using GHIElectronics.TinyCLR.Devices.Gpio;
+using GHIElectronics.TinyCLR.Devices.Spi;
+
+namespace GHIElectronics.TinyCLR.Drivers.Sitronix.ST7735 {
+    public enum ColorFormat {
+        BGR12bit444,
```
Comment (Arke64) [inline Sitronix/GHIElectronics.TinyCLR.Drivers.Sitronix.ST7735/ST7735.cs]: Bgr, not BGR

File: Sitronix/GHIElectronics.TinyCLR.Drivers.Sitronix.ST7735/ST7735.cs [hunk 2]
```diff
@@ -1,5 +1,222 @@
-﻿namespace GHIElectronics.TinyCLR.Drivers.Sitronix.ST7735 {
+﻿using System.Threading;
+
```
Comment (Arke64) [inline Sitronix/GHIElectronics.TinyCLR.Drivers.Sitronix.ST7735/ST7735.cs]: Remove newline

File: Sitronix/GHIElectronics.TinyCLR.Drivers.Sitronix.ST7735/ST7735.cs [hunk 3]
```diff
@@ -1,5 +1,222 @@
...
+    }
+
     public class ST7735 {
+        private const byte ST7735_MADCTL = 0x36;
+        private const byte MADCTL_MY = 0x80;
+        private const byte MADCTL_MX = 0x40;
+        private const byte MADCTL_MV = 0x20;
+        private const byte MADCTL_BGR = 0x08;
+
+        private readonly SpiController spiBusContoller;
```
Comment (Arke64) [inline Sitronix/GHIElectronics.TinyCLR.Drivers.Sitronix.ST7735/ST7735.cs]: Don't store controller

File: Sitronix/GHIElectronics.TinyCLR.Drivers.Sitronix.ST7735/ST7735.cs [hunk 4]
```diff
@@ -1,5 +1,222 @@
...
+        private readonly SpiDevice spiBus;
+
+        private readonly GpioPin resetPin;
+        private readonly GpioPin controlPin;
+
+        private readonly byte[] singleByteBuffer;
```
Comment (Arke64) [inline Sitronix/GHIElectronics.TinyCLR.Drivers.Sitronix.ST7735/ST7735.cs]: The name for something like this that we usually follow is `buffer1`, `buffer2`, `buffer4`, etc

File: Sitronix/GHIElectronics.TinyCLR.Drivers.Sitronix.ST7735/ST7735.cs [hunk 5]
```diff
@@ -1,5 +1,222 @@
...
+        private int drawWindowY = 0;
+        private int drawWindowWidth = maxWidth;
+        private int drawWindowHeight = maxHeight;
+
+        private int bitsPerPixel = 16;
+
+        /// <summary>
+        /// The width of the display in pixels.
+        /// </summary>
+        public const int maxWidth = 160;
```
Comment (Arke64) [inline Sitronix/GHIElectronics.TinyCLR.Drivers.Sitronix.ST7735/ST7735.cs]: MaxWidth

File: Sitronix/GHIElectronics.TinyCLR.Drivers.Sitronix.ST7735/ST7735.cs [hunk 6]
```diff
@@ -1,5 +1,222 @@
...
+
+        /// <summary>
+        /// The height of the display in pixels.
+        /// </summary>
+        public const int maxHeight = 128;
```
Comment (Arke64) [inline Sitronix/GHIElectronics.TinyCLR.Drivers.Sitronix.ST7735/ST7735.cs]: MaxHeight

File: Sitronix/GHIElectronics.TinyCLR.Drivers.Sitronix.ST7735/ST7735.cs [hunk 7]
```diff
@@ -1,5 +1,222 @@
...
+
+        public ST7735(int resetPin, int controlPin, int chipSelect, string spiId) {
+            this.singleByteBuffer = new byte[1];
```
Comment (Arke64) [inline Sitronix/GHIElectronics.TinyCLR.Drivers.Sitronix.ST7735/ST7735.cs]: Initialize these inline above in the var decl

File: Sitronix/GHIElectronics.TinyCLR.Drivers.Sitronix.ST7735/ST7735.cs [hunk 8]
```diff
@@ -1,5 +1,222 @@
...
+            this.twoByteBuffer = new byte[2];
+            this.wordBuffer = new byte[4];
+
+            var GPIO = GpioController.GetDefault();
```
Comment (Arke64) [inline Sitronix/GHIElectronics.TinyCLR.Drivers.Sitronix.ST7735/ST7735.cs]: gpio, not GPIO

File: Sitronix/GHIElectronics.TinyCLR.Drivers.Sitronix.ST7735/ST7735.cs [hunk 9]
```diff
@@ -1,5 +1,222 @@
...
+
+            this.controlPin = GPIO.OpenPin(controlPin);
+            this.controlPin.SetDriveMode(GpioPinDriveMode.Output);
+
+            this.resetPin = GPIO.OpenPin(resetPin);
+            this.resetPin.SetDriveMode(GpioPinDriveMode.Output);
+
+            var spiSettings = new SpiConnectionSettings(chipSelect) {
```
Comment (Arke64) [inline Sitronix/GHIElectronics.TinyCLR.Drivers.Sitronix.ST7735/ST7735.cs]: Inline to the call to `GetDevice`

File: Sitronix/GHIElectronics.TinyCLR.Drivers.Sitronix.ST7735/ST7735.cs [hunk 10]
```diff
@@ -1,5 +1,222 @@
...
+            };
+
+            this.spiBusContoller = SpiController.FromName(spiId);
+            this.spiBus = this.spiBusContoller.GetDevice(spiSettings);
+
+            this.Reset();
+            this.InitializeST7735();
+        }
+
+        private void InitializeST7735() {
```
Comment (Arke64) [inline Sitronix/GHIElectronics.TinyCLR.Drivers.Sitronix.ST7735/ST7735.cs]: Just `Initialize`

File: Sitronix/GHIElectronics.TinyCLR.Drivers.Sitronix.ST7735/ST7735.cs [hunk 11]
```diff
@@ -1,5 +1,222 @@
...
+                    this.SendData(0x03);
+                    break;
+                case ColorFormat.BGR16bit565: // 65k colors mode
+                    this.bitsPerPixel = 16;
+                    this.SendCommand(0x3A); 
+                    this.SendData(0x05);
+                    break;
+            }
+        }
+        public void SetDrawWindow(int x, int y, int width, int height) {
```
Comment (Arke64) [inline Sitronix/GHIElectronics.TinyCLR.Drivers.Sitronix.ST7735/ST7735.cs]: Need newline

File: Sitronix/GHIElectronics.TinyCLR.Drivers.Sitronix.ST7735/ST7735.cs [hunk 12]
```diff
@@ -1,5 +1,222 @@
...
+        private readonly byte[] singleByteBuffer;
+        private readonly byte[] twoByteBuffer;
+        private readonly byte[] wordBuffer;
+
+        private int drawWindowX = 0;
+        private int drawWindowY = 0;
+        private int drawWindowWidth = maxWidth;
+        private int drawWindowHeight = maxHeight;
+
+        private int bitsPerPixel = 16;
```
Comment (Arke64) [inline Sitronix/GHIElectronics.TinyCLR.Drivers.Sitronix.ST7735/ST7735.cs]: Make this into a public property, private setter. And change the type into the above `ColorFormat` enum

File: Sitronix/GHIElectronics.TinyCLR.Drivers.Sitronix.ST7735/ST7735.cs [hunk 13]
```diff
@@ -1,5 +1,222 @@
...
+        private readonly GpioPin resetPin;
+        private readonly GpioPin controlPin;
+
+        private readonly byte[] singleByteBuffer;
+        private readonly byte[] twoByteBuffer;
+        private readonly byte[] wordBuffer;
+
+        private int drawWindowX = 0;
+        private int drawWindowY = 0;
+        private int drawWindowWidth = maxWidth;
```
Comment (Arke64) [inline Sitronix/GHIElectronics.TinyCLR.Drivers.Sitronix.ST7735/ST7735.cs]: Don't set to maxWidth or maxHeight here, do that in the constructor with a call to SetDrawWindow.

File: Sitronix/GHIElectronics.TinyCLR.Drivers.Sitronix.ST7735/ST7735.cs (hunk 4 above)
Comment (Experiment-626) [reply Sitronix/GHIElectronics.TinyCLR.Drivers.Sitronix.ST7735/ST7735.cs]: I will change it to buffer1 but I think singleByteBuffer, while wordier, is more concise.

**Pull request merged**